doc/rules/php_unit/php_unit_data_provider_return_type.rst

```diff
@@ -0,0 +1,50 @@
+===========================================
+Rule ``php_unit_data_provider_return_type``
```

Comment:
should we add to our ruleset?

is it that simply we (Fixer team) wants this change, or it's expected interface of provider method by PHPUnit ? (depends on this, different ruleset target)

Reply:
Yes, I agreed with @Wirone that he will raise the PR after this one is merged (so we can focus here on the fixer here).

That way he will not have to see the diff there will be in data providers 😄

Reply:
would be nice to see results of the fixer before merging it. maybe side PR that we apply changes of this fixer, but that side PR to be merged first? then, in this PR you can already enable it and we see it running in CI for our codebase

Reply:
sure, here it is: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7072

Reply:
Diff is smaller than I initially thought, because I expected that it will also replace arrays with yields 😆. I believe there was rule for that somewhere.. 🤔.

Reply:
> should we add to our ruleset?
> 
> is it that simply we (Fixer team) wants this change, or it's expected interface of provider method by PHPUnit ? (depends on this, different ruleset target)

@kubawerlos ^^

Reply:
https://docs.phpunit.de/en/10.2/writing-tests-for-phpunit.html#data-providers

```
A data provider [...] must either return an array of arrays or an object that implements the Iterator interface.
```

so `dataprovider` may return `array` and not only `iterable` it is not a hard requirement by PHPUnit

Reply:
I'd still stay with the claim that it is `iterable`: https://github.com/sebastianbergmann/phpunit/blob/10.2.2/src/Metadata/Api/DataProvider.php#L174-L194

Reply:
I believe @SpacePossum's comment was only about _expectation_ @keradus asked before. There isn't preferred way, PHPUnit just supports both. So if anything, it's not strict requirement, just convention to choose.

Reply:
Hard to tell if it is expected, I'd say it is a matter of personal taste, `iterable` is the contract. I can imagine a team of developers that agree to use `array` as much as one agreeing to use `\Generator`.

Reply:
> https://docs.phpunit.de/en/10.2/writing-tests-for-phpunit.html#data-providers
> 
> ```
> A data provider [...] must either return an array of arrays or an object that implements the Iterator interface.
> ```

@SpacePossum It is outdated now 😉

Reply:
> It must return [iterable](https://www.php.net/manual/en/language.types.iterable.php), either an array of arrays or an object that implements the Traversable interface.


hmm... let's quote it in Fixer description and put it in PHPUnit ruleset?

Reply:
PHPUnit rulesets are for migrating to a higher PHPUnit version and this one is related to PHP, not PHPUnit version.

Don't we want it in `PhpCsFixer:risky`?

Reply:
@kubawerlos was `array|iterable` support in PHPUnit from day one or it's like that since specific version? Because maybe it's possible to define proper target set.

IMHO it could be added to `PhpCsFixer:risky`.

Reply:
It was always that return type, PHPUnit 6 is the first one to support PHP 7+ only,.

Reply:
but also PHPunit updated their docs to explicitly state `iterable` ? maybe we can take version for which they claim so?

Reply:
ok, I'm also happy with PHPUnit 6, but maybe worth a comment in code about this decision.

Reply:
> It was always that return type, PHPUnit 6 is the first one to support PHP 7+ only,.


if someone would write code in PHP7 , but run PHPUnit 5 (which allow PHP7), it would still work for them, right?
imho as low as possible [so even before PHPUnit 5], or matching the version PHPUnit updated their docs - PHPUnit 10.

Reply:
Let's make one step backward:

All the `@PHPUnitXXMigration` sets are to migrate from one PHPUnit version to another - higher - one, due to some classes changing name from underscores to namespaced, some assertions deprecated/removed/renamed/refactored or mocking changes.

This one is not related to PHPUnit **MIGRATION** more than (random fixers here) `PhpUnitInternalClassFixer` or `PhpUnitTestAnnotationFixer` - they are nice to have, but not related to any PHPUnit version.

_(So I've removed the fixer from `@PHPUnit60Migration:risky` set)_

Reply:
Maybe we need something like `@PHPUnitGoodPractices` 🤔?

Reply:
I see a change on PHPUnit side.

PHPunit -9 was asking to have dataproviders returning array.
PHPUnit 10 started asking to have them returning iterable.

Even if code-level spec didn't changed, expectations named in docs did.

----

Maybe we can be lucky and get attention from @sebastianbergmann - Sebastian, if you noticed this question, what would be your take on it? Would you recommend to migrate code for dataproviders to return iterable explicitly?

Reply:
It still would be "recommended", not required to migrate, like I guess using `function testFoo() {}` is recommended over `/** @test */ function foo() {}`.

As long as it is not required to migrate to a higher PHPUnit version I'd say let's not add it to migrations.

And definitely, this is something we can make in separate PR 😉

Reply:
I am not aware of any change in PHPUnit 10 vs. PHPUnit 9 with regard to what a data provider is expected to return. PHPUnit just expects that whatever is returned can be iterated over and that each iteration yields an array.